### ISSUE TITLE
Ncurses Extended Color Pairs incorrectly display as copies of lower 256 color pairs

### ISSUE THREAD

**Issue by cmang** [python/cpython#119138]
# Bug report

### Bug description:

When using ncurses extended color pairs, introduced in Python 3.10 in PR #[17536](https://github.com/python/cpython/pull/17536) and issue #81163, the color pairs past 256 don't seem to work. When accessed, pairs 256+ repeat the behavior of pairs 0-255 instead of the colors set by curses.init_pair().

In my use case, curses.has_extended_color_support() == True and curses.COLORS == 256.

I can use curses.init_pair() to initialize color pairs past 255. It lets me set them without any error. But when I go to display the colors, color pair 256 displays the same as color pair 0, 257 displays the same as 1, and so on.

I created a small test program to reproduce the issue by initializing, displaying, and letting you scroll through the extended color pairs, examining pairs 0-65536: https://gist.github.com/cmang/7d366f677cd067c00e58a9d9c97141c5

The test program initializes 256 * 256 (65536) color pairs, using every FG and BG combination for FG and BG colors 0-255. It then lets you use the left/right arrow keys to scroll through the pairs. It requires an xterm-256color capable terminal. 

In the following screenshot, the test program shows each color pair number in its given color and you can see that the pairs repeat. The expected behavior is that pairs 256-512 should have a different background color set (red), as should the next 256 pairs (green), etc.

<img width="715" alt="Screenshot 2024-05-18 at 4 18 41 AM" src="https://github.com/python/cpython/assets/261501/e7c8fd87-e885-4884-aa86-4bab50ac04a3">

Edit: If you get an error running the test program, make your terminal larger.

@hpjansson @websurfer5 

Thanks!

### CPython versions tested on:

3.10, 3.11, 3.12

### Operating systems tested on:

Linux, macOS

**Comment by hpjansson** [comment]
The `addstr` and `addch` functions take an attribute value, but attributes can only store a color pair in the range 0..255. `COLOR_PAIR()` applies a bitmask so it can be combined with other attributes. I think we need new API to allow users to pass in the color pair separately and use `wattr_set()` internally instead of the legacy `wattrset()`. Ref. `attr(3NCURSES)`.

**Comment by EvanCarroll** [comment]
@hpjansson why do we have to maintain an older api with `wattrset()` if `wattr_set()` can do all the same things, and more. And, we never expose that to the user directly can't we just update the binding to use `wattr_set()` under the hood and fix this?

**Comment by grymmjack** [comment]
Would love to know the status of this!

**Comment by picnixz** [comment]
> why do we have to maintain an older api with wattrset() if wattr_set() can do all the same things,

Because not everyone has access to the new API. So we need to maintain whatever is still being used in the wild.

> I think we need new API to allow users to pass in the color pair separately and use wattr_set() internally instead of the legacy wattrset().

I can think of this. Or we can call this one when we pass a color attribute internally so that we don't have a new public API. I'll need to think about it.

**Comment by EvanCarroll** [comment]
> Because not everyone has access to the new API. So we need to maintain whatever is still being used in the wild.

When you say "new API" here you mean a new enough version of ncurses? If so, can't we just provide different definitions for the same function -- one which uses the new ncurses API and one which uses the old one? That's pretty routine for C. Then if you have the new API, the advanced color functionality just works because `wattr_set()` is called under the hood.

**Comment by picnixz** [comment]
> When you say "new API" here you mean a new enough version of ncurses? 

Yes, according to the docs, if you want to have more than 256 color pairs, you need ncurses ABI 6 *even if you use `wattr_set`*. But maybe we could just make the detection of extended pairs better (namely, if the ABI is < 6 and `wattr_set` does not exist => we say we cannot support extended pairs).

**Comment by cmang** [comment]
>  maybe we could just make the detection of extended pairs better (namely, if the ABI is < 6 and `wattr_set` does not exist => we say we cannot support extended pairs).

I reckon that since the documentation already hides both ABI versions behind one Python interface, then this would be a nice way for reality to reflect the documentation, if possible. 

To me this seems preferable over making a new interface on the Python side, as people are already programming based on the documentation. It's also less work for people using the interface, if a bit more work  behind-the-scenes on the Cpython side.

This approach deviates somewhat from the C ncurses patterns to be more Pythonic, though. I think that is cool if it works, but bad if it doesn't. Right now it doesn't.

Dumb question, but will this have implications for other Python implementations? Are there any non-cpython issues to take into consideration?

**Comment by picnixz** [comment]
> will this have implications for other Python implementations

This is an extension module, entirely optional, and I believe that PyPy and others just ships the ncurses module. The only thing we touch is the curses API calls which are independent of how Python is implemented. Unless other implementations decideed to reimplement the curses module (namely provide their own shared library), then there won't be any impact. If they do, there will be a difference between our and the other implementations but that's their job if they want to backport the new feature. 

> This approach deviates somewhat from the C ncurses patterns to be more Pythonic, though

Actually, that's what we already do internally. When we can, we use new curses ABI calls internally and when we can't then we don't. For instance, for `addstr`:

> The biggest difference is that the Python interface makes things simpler by merging different C functions such as addstr(), mvaddstr(), and mvwaddstr() into a single [addstr()](https://docs.python.org/3/library/curses.html#curses.window.addstr) method.

So we're already unifying multiple kind of calls like this. In addition, we don't expose specific `wattrset` and `wattr_set`, so we're already saying "whatever attribute you want, we'll handle it ourselves".

**Comment by hpjansson** [comment]
> @hpjansson why do we have to maintain an older api with `wattrset()` if `wattr_set()` can do all the same things, and more. And, we never expose that to the user directly can't we just update the binding to use `wattr_set()` under the hood and fix this?

I agree this would be the best solution, as long as it can be done without breaking existing code. I was also worried that it'd break the thin wrapper pattern of the ncurses module, but as @picnixz pointed out, we're already merging calls, so that's probably not an issue.

**Comment by janEntikan** [comment]
I'd also like to be able to use 256 background colors.

**Comment by serhiy-storchaka** [comment]
The pair was passed packed into a `chtype` via `curses.color_pair()`, whose color field only holds about 8 bits, so pairs ≥ 256 wrap to the lower 256.  That is why they set correctly but display wrong.

Python 3.16 adds APIs that pass the color pair as a separate argument instead of packing it, so pairs ≥ 256 work:

* `window.color_set(pair)` / `window.attr_set(attr, pair)` (gh-152219)
* `complexchar(text, attr, pair)` / `complexstr`, accepted by `addch()`/`addstr()` (gh-152233)

So your example can be adapted by setting the pair separately:
```python
win.color_set(pair); win.addstr(text)
```
or packing it with text:
```python
win.addstr(curses.complexstr(text, pair=pair))
```

**Comment by serhiy-storchaka** [comment]
Closing as addressed in 3.16.
